Unify more of xcconfig content across projects

aproskuryakov opened the pull request:
#### e31966cc25b1405462927008cfaa9e06064af8f4
<pre>
Unify more of xcconfig content across projects
<a href="https://bugs.webkit.org/show_bug.cgi?id=249718">https://bugs.webkit.org/show_bug.cgi?id=249718</a>
rdar://103600264

Reviewed by Elliott Williams.

Move common variables from individual projects&apos; xcconfigs to central Configurations directory.

We had significant divergence between variables of the same name in different projects, so
this is more of a refactoring than a mechanical move. Also, did some unrelated cleanup that
was hard to pass.

- Centralized WK_TARGET_{IOS|TVOS|WATCHOS}_VERSION_MAJOR and TARGET_MAC_OS_X_VERSION_MAJOR
variables. The latter had slightly different definitions between xcconfigs, especially in
gtest, but effectively there should be no behavior change.

- Renamed WK_PRIVATE_FRAMEWORKS_DIR to WK_PRIVATE_FRAMEWORK_STUBS_DIR, because the old name
was super misleading.

- Use SDK_DIR other SDKROOT in more places. No real difference, but it&apos;s more idiomatic.

- Removed NEXT_ROOT, which is long obsolete.

- Removed DISABLE_SDK_METADATA_PARSING workaround, which is no longer needed.

- Changed tools projects to build successfully for STP and StagedFrameworks builds. We don&apos;t
need that yet, but it was good to test unification with this, and I&apos;m thinking that we should
look into building tools in these configurations. The biggest difference was creating
symlinks in generate-derived-sources.sh to avoid having spaces in paths passed to `make`.

* Configurations/CommonBase.xcconfig: Added. This is intended to be one file included from
Base.xcconfig in all projects, reducing the risk of missing an important include somewhere.

* Configurations/SDKVariant.xcconfig: Moved a couple variables derived purely from SDK
behavior here, as opposed to WebKit project structure.

* Configurations/WebKitProjectPaths.xcconfig: Added. This contains paths to projects&apos; SDK
content intended to be used in dependent projects. Each project is still responsible for
calculating its SEARCH or INSTALL paths, but their calculation also gets simpler and more
unified using variables from this file.

* Source/WebKit/Configurations/Shim.xcconfig: Removed.
* Tools/DumpRenderTree/mac/Configurations/TestNetscapePlugIn.xcconfig: Removed.
    These weren&apos;t used by any targets any more.

* Source/JavaScriptCore/Configurations/Base.xcconfig:
* Source/JavaScriptCore/Configurations/TestAPI.xcconfig:
* Source/JavaScriptCore/Configurations/ToolExecutable.xcconfig:
* Source/ThirdParty/ANGLE/Configurations/ANGLE-dynamic.xcconfig:
* Source/ThirdParty/ANGLE/Configurations/AngleMetalLib.xcconfig:
* Source/ThirdParty/ANGLE/Configurations/Base.xcconfig:
* Source/ThirdParty/gtest/xcode/Config/General.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/Base.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/opus.xcconfig:
* Source/WTF/Configurations/Base.xcconfig:
* Source/WebCore/Configurations/Base.xcconfig:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Configurations/WebCoreTestSupport.xcconfig:
* Source/WebCore/PAL/Configurations/Base.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/libavif/Configurations/Base.xcconfig:
* Source/WebCore/PAL/libavif/Configurations/libavif.xcconfig:
* Source/WebCore/PAL/libavif/dav1d/Configurations/Base.xcconfig:
* Source/WebCore/PAL/libavif/dav1d/Configurations/dav1d.xcconfig:
* Source/WebGPU/Configurations/Base.xcconfig:
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebInspectorUI/Configurations/Base.xcconfig:
* Source/WebInspectorUI/Configurations/WebInspectorUIFramework.xcconfig:
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/Configurations/BaseXPCService.xcconfig:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Configurations/WebKitSwift.xcconfig:
* Source/WebKit/Configurations/adattributiond.xcconfig:
* Source/WebKit/SwiftOverlay/Configurations/Base.xcconfig:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/mac/Configurations/Base.xcconfig:
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig:
* Source/bmalloc/Configurations/Base.xcconfig:
* Tools/DumpRenderTree/DumpRenderTree.xcodeproj/project.pbxproj:
* Tools/DumpRenderTree/Scripts/generate-derived-sources.sh:
* Tools/DumpRenderTree/mac/Configurations/Base.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/DumpRenderTree.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/DumpRenderTreeLibrary.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/LayoutTestHelper.xcconfig:
* Tools/ImageDiff/cg/Configurations/Base.xcconfig:
* Tools/MiniBrowser/Configurations/Base.xcconfig:
* Tools/MobileMiniBrowser/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Configurations/DebugRelease.xcconfig:
* Tools/TestWebKitAPI/Configurations/InjectedBundle.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestIPC.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestWGSL.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestWTFLibrary.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestWebKitAPI.xcconfig:
* Tools/WebEditingTester/Configurations/Base.xcconfig:
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:
* Tools/WebKitTestRunner/Configurations/BaseTarget.xcconfig:
* Tools/WebKitTestRunner/Configurations/InjectedBundle.xcconfig:
* Tools/WebKitTestRunner/Configurations/WebKitTestRunner.xcconfig:
* Tools/WebKitTestRunner/Scripts/generate-derived-sources.sh:
* Tools/lldb/lldbWebKitTester/Configurations/Base.xcconfig:

Canonical link: <a href="https://commits.webkit.org/258266@main">https://commits.webkit.org/258266@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3851e610be64a7d329f122af8f33c265b600bd6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101397 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34455 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110668 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105377 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11505 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1407 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93791 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/108494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107179 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/8748 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/35288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/23397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/91832 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4166 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/87969 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/1742 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4221 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/1328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29310 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/10316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44397 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/90869 "Built successfully") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5688 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5985 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20299 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -53,55 +51,26 @@ PROFILE_GENERATE_OR_USE_LDFLAGS = $(PROFILE_GENERATE_OR_USE_LDFLAGS_$(ENABLE_LLV
 HEADER_SEARCH_PATHS = PAL ForwardingHeaders /usr/include/libxslt /usr/include/libxml2 "$(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore" "$(BUILT_PRODUCTS_DIR)$(WK_LIBRARY_HEADERS_FOLDER_PATH)" $(WEBKITADDITIONS_HEADER_SEARCH_PATHS) $(ANGLE_HEADER_SEARCH_PATHS) $(LIBWEBRTC_HEADER_SEARCH_PATHS) $(HEADER_SEARCH_PATHS) $(SRCROOT);
 SYSTEM_HEADER_SEARCH_PATHS = $(PROJECT_DIR)/PAL/libavif/include $(SDK_DIR)$(WK_ALTERNATE_WEBKIT_SDK_PATH)$(WK_LIBRARY_HEADERS_FOLDER_PATH) $(inherited);
-LIBRARY_SEARCH_PATHS = $(SDK_DIR)$(WK_LIBRARY_INSTALL_PATH) $(inherited);
+LIBRARY_SEARCH_PATHS = "$(SDK_DIR)$(WEBCORE_LIBRARY_DIR)" $(SDK_DIR)$(WK_LIBRARY_INSTALL_PATH) $(inherited);
```
nit: Quotes should not be necessary here (and if they are, I'm curious to know why!)

aproskuryakov replied:
`WEBCORE_LIBRARY_DIR` can be `/Applications/Safari Technology Preview.app/Contents/Frameworks`, it is where libwebrtc and libANGLE-shared go. This is different from the `WK_LIBRARY_INSTALL_PATH` SDK library path for static libraries.

We do need to make these names more clear. The actual locations don't necessarily make sense either.

emw-apple commented on Source/WebKit/Configurations/BaseTarget.xcconfig:
```diff
@@ -141,6 +74,7 @@ WK_SKIP_INSTALL_WEBPUSHD[sdk=macosx*] = $(WK_OR_$(WK_RELOCATABLE_FRAMEWORKS)_$(W
 WK_SKIP_INSTALL_WEBPUSHD[sdk=macosx11*] = YES;
 WK_SKIP_INSTALL_WEBPUSHD[sdk=macosx12*] = YES;
 
+// FIXME: There is no need for a variable that's always YES. If this ever needs to be disabled, it should be a USE, not a HAVE.
 WK_HAVE_CORE_PREDICTION = YES;
```
Was curious about this setting so I looked up how it's used. As a follow up, I think we can simplify it to just this line in WebKit.xcconfig:

```
WK_CORE_PREDICTION_LDFLAGS = $(WK_MACOS_WEAK_FRAMEWORK) CorePrediction
```

and we'll need to unifdef the six `HAVE(CORE_PREDICTION)` blocks in WebKit.

emw-apple commented on Tools/DumpRenderTree/mac/Configurations/DumpRenderTreeLibrary.xcconfig:
```diff
@@ -28,10 +28,6 @@
 PRODUCT_NAME = DumpRenderTree;
 SKIP_INSTALL = YES;
 
-// Workaround for "WebKit is not available..." error (rdar://91668054)
-DISABLE_SDK_METADATA_PARSING[sdk=appletv*] = YES;
-DISABLE_SDK_METADATA_PARSING[sdk=watch*] = YES;
-
```
Confirmed that while this would break builds for tv/watchOS on Xcode 13.x, we only build embedded WebKit with 14.x.

emw-apple reviewed (approved):
Looks good, and thank you endlessly for the organization of `WebKitProjectPaths.xcconfig`!

aproskuryakov committed:
Unify more of xcconfig content across projects
https://bugs.webkit.org/show_bug.cgi?id=249718
rdar://103600264

Reviewed by Elliott Williams.

Move common variables from individual projects' xcconfigs to central Configurations directory.

We had significant divergence between variables of the same name in different projects, so
this is more of a refactoring than a mechanical move. Also, did some unrelated cleanup that
was hard to pass.

- Centralized WK_TARGET_{IOS|TVOS|WATCHOS}_VERSION_MAJOR and TARGET_MAC_OS_X_VERSION_MAJOR
variables. The latter had slightly different definitions between xcconfigs, especially in
gtest, but effectively there should be no behavior change.

- Renamed WK_PRIVATE_FRAMEWORKS_DIR to WK_PRIVATE_FRAMEWORK_STUBS_DIR, because the old name
was super misleading.

- Use SDK_DIR other SDKROOT in more places. No real difference, but it's more idiomatic.

- Removed NEXT_ROOT, which is long obsolete.

- Removed DISABLE_SDK_METADATA_PARSING workaround, which is no longer needed.

- Changed tools projects to build successfully for STP and StagedFrameworks builds. We don't
need that yet, but it was good to test unification with this, and I'm thinking that we should
look into building tools in these configurations. The biggest difference was creating
symlinks in generate-derived-sources.sh to avoid having spaces in paths passed to `make`.

* Configurations/CommonBase.xcconfig: Added. This is intended to be one file included from
Base.xcconfig in all projects, reducing the risk of missing an important include somewhere.

* Configurations/SDKVariant.xcconfig: Moved a couple variables derived purely from SDK
behavior here, as opposed to WebKit project structure.

* Configurations/WebKitProjectPaths.xcconfig: Added. This contains paths to projects' SDK
content intended to be used in dependent projects. Each project is still responsible for
calculating its SEARCH or INSTALL paths, but their calculation also gets simpler and more
unified using variables from this file.

* Source/WebKit/Configurations/Shim.xcconfig: Removed.
* Tools/DumpRenderTree/mac/Configurations/TestNetscapePlugIn.xcconfig: Removed.
    These weren't used by any targets any more.

* Source/JavaScriptCore/Configurations/Base.xcconfig:
* Source/JavaScriptCore/Configurations/TestAPI.xcconfig:
* Source/JavaScriptCore/Configurations/ToolExecutable.xcconfig:
* Source/ThirdParty/ANGLE/Configurations/ANGLE-dynamic.xcconfig:
* Source/ThirdParty/ANGLE/Configurations/AngleMetalLib.xcconfig:
* Source/ThirdParty/ANGLE/Configurations/Base.xcconfig:
* Source/ThirdParty/gtest/xcode/Config/General.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/Base.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/opus.xcconfig:
* Source/WTF/Configurations/Base.xcconfig:
* Source/WebCore/Configurations/Base.xcconfig:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Configurations/WebCoreTestSupport.xcconfig:
* Source/WebCore/PAL/Configurations/Base.xcconfig:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/libavif/Configurations/Base.xcconfig:
* Source/WebCore/PAL/libavif/Configurations/libavif.xcconfig:
* Source/WebCore/PAL/libavif/dav1d/Configurations/Base.xcconfig:
* Source/WebCore/PAL/libavif/dav1d/Configurations/dav1d.xcconfig:
* Source/WebGPU/Configurations/Base.xcconfig:
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebInspectorUI/Configurations/Base.xcconfig:
* Source/WebInspectorUI/Configurations/WebInspectorUIFramework.xcconfig:
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/Configurations/BaseXPCService.xcconfig:
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/Configurations/WebKitSwift.xcconfig:
* Source/WebKit/Configurations/adattributiond.xcconfig:
* Source/WebKit/SwiftOverlay/Configurations/Base.xcconfig:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/mac/Configurations/Base.xcconfig:
* Source/WebKitLegacy/mac/Configurations/WebKitLegacy.xcconfig:
* Source/bmalloc/Configurations/Base.xcconfig:
* Tools/DumpRenderTree/DumpRenderTree.xcodeproj/project.pbxproj:
* Tools/DumpRenderTree/Scripts/generate-derived-sources.sh:
* Tools/DumpRenderTree/mac/Configurations/Base.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/DumpRenderTree.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/DumpRenderTreeLibrary.xcconfig:
* Tools/DumpRenderTree/mac/Configurations/LayoutTestHelper.xcconfig:
* Tools/ImageDiff/cg/Configurations/Base.xcconfig:
* Tools/MiniBrowser/Configurations/Base.xcconfig:
* Tools/MobileMiniBrowser/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Configurations/Base.xcconfig:
* Tools/TestWebKitAPI/Configurations/DebugRelease.xcconfig:
* Tools/TestWebKitAPI/Configurations/InjectedBundle.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestIPC.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestWGSL.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestWTFLibrary.xcconfig:
* Tools/TestWebKitAPI/Configurations/TestWebKitAPI.xcconfig:
* Tools/WebEditingTester/Configurations/Base.xcconfig:
* Tools/WebKitTestRunner/Configurations/Base.xcconfig:
* Tools/WebKitTestRunner/Configurations/BaseTarget.xcconfig:
* Tools/WebKitTestRunner/Configurations/InjectedBundle.xcconfig:
* Tools/WebKitTestRunner/Configurations/WebKitTestRunner.xcconfig:
* Tools/WebKitTestRunner/Scripts/generate-derived-sources.sh:
* Tools/lldb/lldbWebKitTester/Configurations/Base.xcconfig:

Canonical link: https://commits.webkit.org/258266@main
